feat : review report api development

chadongmin opened the pull request:
Resolves #{이슈-번호}
<!--
#214 
-->

# 해결하려는 문제가 무엇인가요?

- 리뷰 신고 기능을 구현합니다.

# 어떻게 해결했나요?

- ReviewFacade를 의존하도록 구현했습니다.


# 어떤 부분에 집중하여 리뷰해야 할까요?

-

## 참고 자료

-

---

# RCA 룰

r: 꼭 반영해 주세요. 적극적으로 고려해 주세요. (Request changes)  
c: 웬만하면 반영해 주세요. (Comment)  
a: 반영해도 좋고 넘어가도 좋습니다. 그냥 사소한 의견입니다. (Approve)

chadongmin committed:
refactor : 기존 엔티티 수정 및 repository 생성

chadongmin committed:
feat : 리뷰 신고 기능 구현

chadongmin committed:
feat : 리뷰 신고 테이블에 ip_address 컬럼 추가 / 사용자 IP 수집 유틸리티 클래스 정의 및 적용

chadongmin committed:
refactor : 검증로직 및 리뷰 자동 Disable처리 로직 추가

- 사용자는 하나의 리뷰는 한번만 신고 가능
- 같은 IP는 하나의 신고로 간주함
- 리뷰 5회 이상 신고 누적시 자동 DISABLE 처리

Whale0928 commented on src/main/java/app/bottlenote/review/domain/Review.java:
```diff
@@ -188,4 +188,12 @@ public void addReply(ReviewReply reply) {
 		this.registerEvent(replyRegistryPublish(this.id, this.userId, reply.getContent()));
 	}
+
+	private static final long BLOCK_THRESHOLD = 5L;
```
이걸 여기에 선언하면 도메인 필드가 되는건대 굳이 static final일 필요성이 없네요.
메소드 내부에서만 관리해도 충분할것같네요.

chadongmin replied:
넵 반영하겠습니다

Whale0928 commented on src/main/java/app/bottlenote/review/service/ReviewFacade.java:
```diff
@@ -1,7 +1,12 @@
+
+	boolean isExistReview(Long reviewId);
+
+	Review getReview(Long reviewId);
```
퍼사드는 엔티티 자체를 노출하면 절대 안됩니다 이러면 퍼사드를 쓰는 이유가 없어요.

chadongmin replied:
Review 엔티티 내부의 updateStatus 와 같은 메서드들을 Facade에 위치시키는게 바로 호출하는게 나은가요??

Whale0928 replied:
`updateStatus` = 명령적 메소드 `requestReviewBlock` = 요청적 메소드

Whale0928 commented on src/main/java/app/bottlenote/support/report/domain/ReviewReport.java:
```diff
@@ -2,44 +2,57 @@
+
+	@Comment("신고자 IP주소")
+	@Column(name = "ip_address", nullable = false)
+	private String ipAddress;
```
Custom Ip Address 타입을 선언해도 깔끔하겟네요

chadongmin replied:
이부분은 추후에 Embeddable 적용할게요

Whale0928 commented on src/main/java/app/bottlenote/support/report/service/ReviewReportService.java:
```diff
@@ -0,0 +1,68 @@
+			ReviewReport saved = reviewReportRepository.save(reviewReport);
+			blockReviewIfNecessary(saved.getReviewId());
+
+		} catch (Exception e) {
```
Exception e은 왜 필요한건가요? 의미없는 예외 캐치같은데

chadongmin replied:
수정했습니다.

Whale0928 commented on src/main/java/app/bottlenote/support/report/service/ReviewReportService.java:
```diff
@@ -0,0 +1,68 @@
+		if (!reviewFacade.isExistReview(reviewReportRequest.reportReviewId())) {
+			throw new ReviewException(REVIEW_NOT_FOUND);
+		}
+		if (reviewReportRepository.existsByUserIdAndReviewId(currentUserId, reviewReportRequest.reportReviewId())) {
```
ifpresent로 처리하는게 훨씬 깔끔해보이네요.

chadongmin replied:
ifPresentOrElse 적용했습니다.

Whale0928 reviewed (commented):
퍼사드를 쓰는이유에 대해서 한번 더 생각해보고 수정해보세요.

Whale0928 commented on src/main/java/app/bottlenote/support/report/service/ReviewReportService.java:
```diff
@@ -0,0 +1,68 @@
+	}
+
+	public void blockReviewIfNecessary(Long reviewId) {
+		Long count = reviewReportRepository.countUniqueIpReportsByReviewId(reviewId).orElse(0L);
```
여기서 단순히 IP만 카운트 하는게 맞나요 ?다시한번 생각해보죠

chadongmin committed:
refactor : 코드리뷰 반영 및 리팩토링

- Facade에서 로직을 수행하도록 수정
- if 구문을 줄이고 stream의 ifPresentOrElse를 사용해서 로직 간결하게 수정
- 쿼리 수정

chadongmin committed:
test : 통합테스트 추가
